app/scripts/controllers/metametrics.js

```diff
@@ -0,0 +1,415 @@
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsEventOptions} MetaMetricsEventOptions
+ * @typedef {import('../../../shared/constants/metametrics').SegmentEventPayload} SegmentEventPayload
+ * @typedef {import('../../../shared/constants/metametrics').SegmentInterface} SegmentInterface
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsPagePayload} MetaMetricsPagePayload
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsPageOptions} MetaMetricsPageOptions
+ */
+
+/**
+ * @typedef {Object} MetaMetricsControllerState
+ * @property {?string} metaMetricsId - The user's metaMetricsId that will be
+ *  attached to all non-anonymized event payloads
+ * @property {?boolean} participateInMetaMetrics - The user's preference for
+ *  participating in the MetaMetrics analytics program. This setting controls
+ *  whether or not events are tracked
+ * @property {number} metaMetricsSendCount - How many send transactions have
+ *  been tracked through this controller. Used to prevent attaching sensitive
+ *  data that can be traced through on chain data.
+ */
+
+export default class MetaMetricsController {
+  /**
+   * @param {boolean} isDevOrTestEnvironment - is the extension running in test
+   *  of dev mode? This will impact how segment is initialized
+   * @param {SegmentInterface} segmentMock - when in test or dev mode use this
+   *  mocked Segment module instead of an instance of Analytics
+   * @param {SegmentInterface} segmentLegacyMock - when in test or dev mode use
+   *  this mocked Segment module instead of an instance of Analytics for legacy
+   *  events
+   * @param {string} segmentWriteKey - the write key for the Segment source to
+   *  track new schema events to
+   * @param {string} segmentLegacyWriteKey - the write key for the Segment
+   *  source to tack legacy events to
+   * @param {string} segmentHost - the host to send events to. The default
+   *  host for Analytics constructor is 'https://api.segment.io'
+   * @param {number} flushAt - the maximum number of events to queue before
+   *  sending to Segment
+   * @param {number} flushInterval - the maximum number of milliseconds to wait
+   *  for before sending queued events
+   * @param {Object} preferencesController - The preferences controller, used
+   *  to access and subscribe to preferences that will be attached to events
+   * @param {Object} networkController - The network controller, used to access
+   *  the current network's chainId and network identifier. These properties
+   *  are attached to events
+   * @param {string} version - The version of the extension
+   * @param {string} environment - The environment the extension is running in
+   * @param {MetaMetricsControllerState} - State to initialized with
+   */
+  constructor({
+    isDevOrTestEnvironment,
+    segmentMock,
+    segmentLegacyMock,
+    segmentWriteKey,
+    segmentLegacyWriteKey,
+    segmentHost,
+    flushAt,
+    flushInterval,
```

Comment:
Yep! In the shared controller proposal, I suggest that we drop the usage of `analytics-node` and `analytics-react-native` and instead use the segment API to send events. This would allow for a truly shared implementation that aligns across both platforms. Still, I'd like us to explicitly pass the fetch implementation in. Even if this were to land in extension your suggestion makes a ton of sense.

Reply:
resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1